src/main/scala/is/hail/expr/ir/IR.scala

```diff
@@ -0,0 +1,74 @@
+package is.hail.expr.ir
+
+import is.hail.asm4s.TypeInfo
+
+import scala.language.existentials
+
+sealed case class PrimitiveOp(x: Long)
+
+sealed trait IR
```

Comment:
1. Should I be able to get the IR type?  Will it recurse every time to compute it?

2. I'd like this to plug into the term rewriter so I can use IR nodes inside the MatrixAST/KeyTableAST nodes.  Have you looked at that?  It doesn't have to be in the first version but we need to make sure it is straightforward.

Reply:
2. See below comments for beginning of thoughts on how they communicate. One situation I thought through: in the optimizer case for `FilterVariants(..., true)` this doesn't change except to use IR in place of AST. I think the MatrixAST optimizer will need to run in tandem with the IR optimizer. In particular, as we work outwards from a `MatrixRead` step, we may find that certain annotations are constant, we want an IR `Read` of that annotation to also be known to be constant (in fact, we may not want the `Read` at all). If we are explicitly manipulating `RegionValue`s, then the optimizer should carry an abstract `Region` that tracks whatever information we know about a given memory location. The abstract region should follow the same data flow as the run-time region. This allows the IR optimizer to reason about what `Read(rv)` may return.

Reply:
1.  Can you explain more?  When will this be run?  Is the type part of the case class match signature, or is it out of line?  We already have a type inferrer in AST.  Is it necessary to build another one?

2. 

 > I think the MatrixAST optimizer will need to run in tandem with the IR optimizer.

This was what I was getting at.  I think we need rulesets that contain IR rewrites as well as MatrixAST rewrites.  (I assume we should use one of IR or AST for the names of our IRs at some point.)  It seems like that means IR needs to extend NewAST, no?

 > we want an IR Read

You mean a MatrixRead?  Not sure I follow this bit.

Reply:
1. My impression was that the AST would be replaced by the python-representation (which will build an IR), ergo we need an inferencer for the IR. You'd need to run it before compiling (since the types are relevant to compilation). It's part of the match signature, otherwise it's annoying to match on `IR`s, right?

2. Yeah we could make IR extend `NewAST`. I figured every `MatrixAST` would optimize its constituent `IR`s before applying any matrix-level rules, but we can put them all in the same framework too.

IR `Read` meaning a read from a region that was written in a previous matrix transformation.